### ISSUE TITLE
GraphQl returns "Invalid parameter X" on any filter

### ISSUE THREAD

**Issue by ChamperNet** [strapi/strapi#17995]
### Required System information

Node.js version: v18.17.1
YARN version: 1.22.19
Strapi version: 4.13.3
Database: SQLITE
Operating system: Windows 10
Is your project Javascript or Typescript: JavaScript

### Describe the bug

Unable to resolve graphql queries having "filter" property 

### Steps to reproduce the behavior

1. Trying to get data form table products with this query:
```
#import "/apollo/fragments/shared/Technical.graphql"

query products(
  $locale: I18NLocaleCode
  $limit: Int!
  $start: Int!
  $brands: [String!]
  $categories: [String!]
  $slug: String
  $id: ID
  $ids: [ID]
  $notSlug: String
  $notId: ID
  $name: String
  $priceMinimal: Float
  $priceMaximal: Float
  $bestseller: Boolean
  $sale: Boolean
  $novelty: Boolean
) {
  products(
    locale: $locale
    pagination: { start: $start, limit: $limit }
    sort: "createdAt:desc"
    filters: {
      id: { eq: $id, in: $ids, ne: $notId }
      slug: { eq: $slug, ne: $notSlug }
      name: { containsi: $name }
      brand: { slug: { in: $brands } }
      categories: { slug: { in: $categories } }
      price: { between: [$priceMinimal, $priceMaximal] }
      active: { eq: true }
      technical: {
        novelty: { eq: $novelty }
        sale: { eq: $sale }
        bestseller: { eq: $bestseller }
      }
    }
  ) {
    data {
      id
      attributes {
        name
        slug
        status
        price
        priceSale
        gallery {
          data {
            attributes {
              formats
              url
              caption
              alternativeText
            }
          }
        }
        picture {
          data {
            attributes {
              formats
              url
              caption
              alternativeText
            }
          }
        }
        brand {
          data {
            attributes {
              name
              slug
            }
          }
        }
        categories {
          data {
            id
            attributes {
              name
              slug
            }
          }
        }
        technical {
          ...Technical
        }
      }
    }
    meta {
      pagination {
        total
        page
        pageCount
        pageSize
      }
    }
  }
}
```

2. Sending params in my request:

```
{
    "limit": 12,
    "start": 0,
    "locale": "en",
    "priceMinimal": 0,
    "priceMaximal": 99999
} 
```

3. Getting an error:

```
"message": "Invalid parameter id",
```

### Expected behavior

Receiving an answer from the server which contains data[] with product information.

### Additional context

Then I remove filter { } from the query - everything works fine.

schema.graphql contains the needed fields of course:

```
input ProductFiltersInput {
    active: BooleanFilterInput
    and: [ProductFiltersInput]
    announcement: StringFilterInput
    brand: BrandFiltersInput
    categories: CategoryFiltersInput
    createdAt: DateTimeFilterInput
    description: StringFilterInput
    features: ComponentSharedFeatureFiltersInput
    id: IDFilterInput
    locale: StringFilterInput
    localizations: ProductFiltersInput
    name: StringFilterInput
    not: ProductFiltersInput
    or: [ProductFiltersInput]
    price: FloatFilterInput
    priceSale: FloatFilterInput
    publishedAt: DateTimeFilterInput
    seo: ComponentSharedSeoFiltersInput
    sitemap_exclude: BooleanFilterInput
    slug: StringFilterInput
    status: StringFilterInput
    technical: ComponentSharedTechnicalFiltersInput
    updatedAt: DateTimeFilterInput
}
```

**Comment by mydesweb** [comment]
I also have this issue with Strapi version 4.13.4 , I use $notIn filter and provide an empty array

**Comment by bl96** [comment]
Encountering same issue with filters ($in) and sort

**Comment by millsoft** [comment]
we are also having the same issue. We use 4.13.2.
For us, it is a critical bug because pages, that are using the filtering, are not working now.

**Comment by thommorais** [comment]
Same!

**Comment by omerkim** [comment]
I guess it's the same bug as mine.
https://github.com/strapi/strapi/issues/17945
Strapi needs to solve this

**Comment by rafikmoreira** [comment]
Hello, everyone. I'm also experiencing issues with the filters. If I include an optional variable but don't send it to the query, it returns as an invalid value. However, in certain situations, such as during the initial rendering of a page, I don't want to apply all the filters. This used to work correctly before the update.

**Comment by omerkim** [comment]
@derrickmehaffy Are you guys going to solve this?
The issue starts to be critical.

**Comment by ChrisCrossCrash** [comment]
I haven't been able to update Strapi to the latest version for several weeks now because of multiple critical bugs in the updates that break basic functionality such as this. I'm starting to get worried about the future of Strapi. Perhaps it would be a good idea to have some kind of intervention to get these issues fixed. Maybe a freeze on all new feature development until the most critical issues are resolved.

**Comment by galago** [comment]
Could you say something if you are working on it, so many people have this issue. It's a critical bug

**Comment by djacquemard** [comment]
Hello !

Same issue here, all our filtered listing are broken with latest version (4.13.7). So we are stuck to 4.12.1 for several weeks.

I hope this critical issue to get solved in priority 🙏

**Comment by innerdvations** [comment]
> Could you say something if you are working on it, so many people have this issue. It's a critical bug

[Working on it right now](https://github.com/strapi/strapi/pull/18168). I don't think we will be able to make it into the release tomorrow (though I will try) but it should certainly be in next week's release!

**Comment by derrickmehaffy** [comment]
Escalated due to internal tid 4884

**Comment by derrickmehaffy** [comment]
Jumping it up to critical based on internal discussion

**Comment by innerdvations** [comment]
The reason for this is that queries are being sent with optional variables being used in query params, for example, `filters: { id: { eq: $id } }` becomes `filters: { id: { } }` when id is not sent, which is not a valid query. Newer versions of Strapi throw errors on invalid queries to help catch security issues and other bugs in user code. But in this particular instance, params that are optionally not sent, I agree it should work.

We currently have two solutions, both of which should allow your queries to work again, we are just discussing which one is the best to go with. It would be great if anyone having this problem could test them to see if it resolves your issues.

1) We can [allow empty object params in only GraphQL](https://github.com/strapi/strapi/pull/18168), which would introduce a subtle difference between the rest API and graphql API
2) We can [allow them everywhere](https://github.com/strapi/strapi/pull/18180)

We're currently deciding the best path and will try to release a fix as soon as we're sure it is the right one and will not introduce any other issues.

**Comment by ChamperNet** [comment]
> The reason for this is that queries are being sent with optional variables being used in query params, for example, `filters: { id: { eq: $id } }` becomes `filters: { id: { } }` when id is not sent, which is not a valid query. Newer versions of Strapi throw errors on invalid queries to help catch security issues and other bugs in user code. But in this particular instance, params that are optionally not sent, I agree it should work as an exception.
> 
> We currently have two solutions, both of which should allow your queries to work again, we are just discussing which one is the best to go with. It would be great if anyone having this problem could test them to see if it resolves your issues.
> 
> 1. We can [allow empty object params in only GraphQL](https://github.com/strapi/strapi/pull/18168), which would introduce a subtle difference between the rest API and graphql API
> 2. We can [allow them everywhere](https://github.com/strapi/strapi/pull/18180)
> 
> We're currently deciding the best path and will try to release a fix as soon as we're sure it is the right one and will not introduce any other issues.

My opinion is allowing empty object params both in GraphQL and REST API would be the right solution.

Great that you finally paid attention to that bug. Hope you'll manage this ASAP.

**Comment by Elpichichi** [comment]
I don't know if you are aware of this, but the issue is still remaining with Boolean filters.
```
query getPrintrollMeasurement($id: ID!, $rubOffFilter: Boolean) {
  printrollMeasurement(id: $id) {
    data {
      id
      attributes {
        printrollctx(
          pagination: { pageSize: 20 }
          filters: { rubOff: { notNull: $rubOffFilter } }
        ) {
          id
          rubOff
        }
      }
    }
  }
}
``` 
In our case, we are using the notNull filter and the answer is still:
`"message": "Invalid parameter rubOffFilter",`
After we remove the filter, everything is working fine.
We are still using v4.12.1 because in v4.14.4 pages, that are using this filter, are not working now.

**Comment by innerdvations** [comment]
I'll reopen this since it appears only the original test case was solved but there is still something to solve here.

**Comment by Kricket** [comment]
This issue is also a blocker for us...when can we expect a fix?

**Comment by alex-del-var** [comment]
Hello,
We are also waiting for this fix.
Currently we are still blocked in v4.7 because of this [bug](https://github.com/strapi/strapi/issues/16170) fixed in v4.12 but we cannot upgrade to v4.12 because of this issue. 
Thanks in advance

**Comment by innerdvations** [comment]
I should have time to work on a fix this week, once I have a PR available I will link it here to keep everyone updated.

**Comment by ChamperNet** [comment]
> I should have time to work on a fix this week, once I have a PR available I will link it here to keep everyone updated.

Thank you. 
If you would have some troubles with managing this issue - give me a sign. I'll help if been able.

**Comment by innerdvations** [comment]
#18647 should fix it. I need to confirm the issue is only present with components and not with relations or dynamic zones, and then we can merge :)

**Comment by gregleeper** [comment]
I saw this was merged to main. Do we have an eta no when it will be fixed in a versioned release?

**Comment by Convly** [comment]
> I saw this was merged to main. Do we have an eta no when it will be fixed in a versioned release?

Hey, we have releases coming out every Wednesday, this fix should have been shipped in 4.15.2 (two days ago)

**Comment by EverettQuebral** [comment]
It seems that problem still exist on 4.15.4

**Comment by galago** [comment]
I wonder why this error reoccurs over and over again. Many people affected by it too. :(

**Comment by JannikZed** [comment]
we are on 4.15.5 - don't know exactly, if our problem is the same as this one:
a query with filter (slug is just a string)
```
query GetMCHCity($slug: StringFilterInput, $locale: I18NLocaleCode) {
  mchCities(filters: { slug: $slug }, locale: $locale) {
    data {
      id
      attributes {
        ...MCHCityDetail
        
      }
    }
  }
}
```
results in a Typescript type `StringFilterInput` results in Typescript complaining `Type 'string' is not assignable to type 'StringFilterInput'`

**Comment by galago** [comment]
That's the same problem I have

> we are on 4.15.5 - don't know exactly, if our problem is the same as this one: a query with filter (slug is just a string)
> 
> ```
> query GetMCHCity($slug: StringFilterInput, $locale: I18NLocaleCode) {
>   mchCities(filters: { slug: $slug }, locale: $locale) {
>     data {
>       id
>       attributes {
>         ...MCHCityDetail
>         
>       }
>     }
>   }
> }
> ```
> 
> results in a Typescript type `StringFilterInput` results in Typescript complaining `Type 'string' is not assignable to type 'StringFilterInput'`

**Comment by alicanso** [comment]
+1

**Comment by 4li-ra** [comment]
+1
Strapi 4.17.1 Invalid parameter on Relations filter

**Comment by derrickmehaffy** [comment]
Reopening due to more reports

**Comment by ChamperNet** [comment]
Do you happen to have any updates about that issue?

**Comment by jerony-mo** [comment]
Any updates on this error please? I believe this is quite critical.

Filtering using a field that does exist only in the "child" components is still throwing the same error.

**Comment by arpadgabor** [comment]
This issue seems to affect the REST API's as well.

**Comment by thoranrion** [comment]
+1 
Strapi 4.20.2

**Comment by m-farahmand** [comment]
this issue happened in rest API for me on every relations without  mapped by other side of collection
Strapi version v4.15.5-alpha.5

**Comment by cacuba** [comment]
Still happening on v4.24.0

**Comment by innerdvations** [comment]
@derrickmehaffy I believe this issue needs to be closed and re-opened with the remaining issue(s), but I'll let you decide and handle it if you want 😆

So far there seem to be 4 issues here, 2 of which have been solved, and it's very messy trying to sift through them.

1. ([Solved](https://github.com/strapi/strapi/pull/18180)) The issue in the initial post, is that queries such as
```
filters: {
      id: { eq: $id, in: $ids, ne: $notId }
}
 ```
when no fields were passed in, became an object like
```
{ filters: { id: {} } }
```

and because `{}` is not a valid id, it was throwing an error, so we added an exception to allow that format of query to work since a lot of users were using that pattern for queries.

2. ([Solved](https://github.com/strapi/strapi/pull/18647)) An issue with nested components not being validated incorrectly

3. Add back the feature of "Allow read permissions on related data on a content type that a user has access for even if they do not have access to the related content type itself"

Context: Previously, about a year ago, a security issue was fixed where related content type data could be leaked, basically the same as if right now you enable permissions on the related content type -- that was the cause of all three "issues" being addressed here, the first two of which were actual bugs in the implementation, while the last being that it revealed this flaw / missing feature.

4. Possibly a typescript issue?
